[Photoshop] Split to CMYK first draft ready with action set

AlexBeers3 opened the pull request:
Cleaned up and finished first draft of split_to_CMYK.jsx.  Script splits CMYK channels into new layers and runs an action set to add adjustment layers.  Based on research adding adjustment layers through a basic javascript is not possible and would need a Listener Event plugin script to achieve.  Wonder if you think its worth it?  Let me know what you think of the set up and layout and if it can be more streamlined as well.

AlexBeers3 committed:
add copy / paste channel to corresponding layer

AlexBeers3 committed:
add action for adj layers for script

AlexBeers3 committed:
add comments to split_to_CMYK.jsx

Kaktug committed:
Fix Line Endings

Kaktug commented on split_to_CMYK.jsx:
```diff
@@ -1 +1,83 @@
+var doc = app.activeDocument;
+
+//switch color mode to get 4 channels
+var CMYK = doc.changeMode(ChangeMode.CMYK);
```
I don't think you need to define this action as a variable here -- doesn't look like you reference this `CMYK` variable elsewhere in the script.

If you're just forcing the document to take on CMYK colormode you should be able to change this line to:
`doc.changeMode(changeMode(ChangeMode.CMYK);`

AlexBeers3 replied:
Correct, that variable isn't used any where else so it makes sense to remove.  Removed and added doc.changeMode(ChangeMode.CMYK);  instead

Kaktug commented on split_to_CMYK.jsx:
```diff
@@ -1 +1,83 @@
+      var layerColor = ["Cyan", "Magenta", "Yellow", "Black"]
+      var doc = activeDocument;
+      var layers = doc.artLayers;
+      var cmyk = [
```
Be careful with variable names.  in this case you're getting away with defining `var CMYK` on line 4 and `var cmyk` here on line 13 because Javascript happens to be a case-sensitive language.

I would suggest changing the instance on this line to something like `var cmyk_layers` which is a little more descriptive.

Kaktug commented on split_to_CMYK.jsx:
```diff
@@ -1 +1,83 @@
+  try {
+    if (app.documents.length) {
+      var layerColor = ["Cyan", "Magenta", "Yellow", "Black"]
+      var doc = activeDocument;
```
Double check this one -- I think you may want this to read `app.activeDocument` instead of just `activeDocument`.

AlexBeers3 replied:
removed line 11 completely as it is already a global variable and redundant.  Script still runs as expected

Kaktug commented on split_to_CMYK.jsx:
```diff
@@ -1 +1,83 @@
+      var doc = activeDocument;
+      var layers = doc.artLayers;
+      var cmyk = [
+        layerA = layers.add(),
```
You may want to name these like so for clarity:
- layerC =
- layerM =
- layerY =
- layerK =

Kaktug commented on split_to_CMYK.jsx:
```diff
@@ -1 +1,83 @@
+        layerB = layers.add(),
+        layerC = layers.add(),
+        layerD = layers.add()
+      ];
```
You may want to name these like so for clarity:
    * layerC =
    * layerM =
    * layerY =
    * layerK =

Kaktug commented on split_to_CMYK.jsx:
```diff
@@ -1 +1,83 @@
+
+      if (layerColor) {
+        cmyk[0].name = layerColor[3];
+        cmyk[0].blendMode = BlendMode.NORMAL
```
Be careful with your code styling -- some lines look like they're ending in semi-colons while others like this one aren't.  Not functionally/technically wrong, but for consistency's sake it's better to choose one method over the other.  (Personally prefer to include semi-colons wherever they should exist to prevent any misinterpretation by the javascript engine inside the ES toolkit)

Kaktug commented on split_to_CMYK.jsx:
```diff
@@ -1 +1,83 @@
+        layerD = layers.add()
+      ];
+
+      if (layerColor) {
```
`if (layerColor) {}` seems like an odd check to me.  From what I can tell you're defining layerColors as a string array up above, so is there ever a case where layerColor would be an empty array or nil?

I think you can simplify this whole method to something like:

```
var doc = app.activeDocument;
var layerC = layers.add();
  layerC.name = 'Cyan';
  layerC.blendMode = BlendMode.NORMAL;
var layerM = layers.add();
  layerM.name = 'Magenta';
  layerM.blendMode = BlendMode.MULTIPLY;

[etc]
```
```

Kaktug commented on split_to_CMYK.jsx:
```diff
@@ -1 +1,83 @@
+
+      }
+
+      doc.activeLayer = newLayer;
```
Do you have the newLayer variable defined anywhere in the script?  I'm not sure if I follow the intention of this step.  I'm assuming you're trying to set the active (highlighted) layer to the first of the newly made CMYK layers?

Kaktug commented on split_to_CMYK.jsx:
```diff
@@ -1 +1,83 @@
+
+      doc.activeLayer = newLayer;
+    }
+  } catch (e) {}
```
For debugging purposes you may want to change this line to:

```
} catch (e) {
  alert(e)
}
```

The `catch` statement right now is consuming any error messages and allowing the script to continue processing.  Based on some of the things I'm seeing in the `makeLayers()` method I think your script is silently failing in some places and not giving you proper notice.

Kaktug commented on split_to_CMYK.jsx:
```diff
@@ -1 +1,83 @@
+makeLayers()
+
+//copies channels into corresponding layers
+function splitCyan() {
```
You could probably combine the `splitCyan, splitMagenta, splitYellow, and splitBlack` functions into a single function that takes some arguments.  For example you could pass the name of the layer you're targeting like this:

```
function splitChannels(targetChannel) {
  [...]
  doc.activeChannels = [doc.channels.getByName(targetChannel)];
  [...]
}
splitChannels('Cyan');
```

Where `targetChannel` is the value passed into the function (in my example, the 'Cyan' string)

Kaktug commented on split_to_CMYK.jsx:
```diff
@@ -1 +1,83 @@
+splitBlack()
+
+//sets color mode back to RGB
+var RGB = doc.changeMode(ChangeMode.RGB);
```
You should be able to just call `doc.changeMode(...)` here w/out assigning the action to a var, see my comment above for the CMYK color mode change.

Kaktug reviewed (changes requested):
Hey man,

Made some comments, overall the structure seems there -- just had some comments about code styling and some suggestions to make the script more fault tolerant & DRYer.

If you have any questions let me know.  Looking forward to seeing the next few commits to this branch / PR.

AlexBeers3 committed:
remove redundent variables and renamed others for clarification

AlexBeers3 committed:
this is only a test

AlexBeers3 committed:
add thats what she said

Kaktug committed:
Kev Sneak Commit

AlexBeers3 committed:
add dummy jsx file

AlexBeers3 committed:
clean up code in jsx and add test.txt

AlexBeers3 committed:
fix test.txt conflict

AlexBeers3 reviewed (commented):
Hey man,  cleaned up the code a bit.  Having an issue with the new splitChannels function.  It pulls the channels and pastes them into layers however they are being pasted as new layers instead of pasting into their corresponding layer.  Going to continue to troubleshoot but figured I upload what I got so far so you can check it out.

AlexBeers3 committed:
change if...else to switch...case statement

AlexBeers3 commented:
cleaned up the code and got it to work.  Changed if...else to switch...case statement.  everything looks to be working great.  Let me know what you think.

Kaktug commented on split_to_CMYK.jsx:
```diff
@@ -1 +1,74 @@
+  			layerC.name = 'Cyan';
+  			layerC.blendMode = BlendMode.MULTIPLY;
+}
+makeLayers();
```
To make your code a little easier to follow & debug you'll want to move this function call for `makeLayers()` down to the spot just above line 65 where you are calling the `splitChannel()` functions.  That way you can see your code execution step-by-step instead of mixed in between function declarations (which, on their own won't do anything until called)

Kaktug commented on split_to_CMYK.jsx:
```diff
@@ -1 +1,74 @@
+
+    doc.activeChannels = [doc.channels.getByName(targetChannel)];
+        function selectCopy() {
+          doc.selection.selectAll();
```
You'll probably want to move this `selectCopy()` function outside of the `splitChannels()` function.  It should be able to live on its own.

Kaktug commented on split_to_CMYK.jsx:
```diff
@@ -1 +1,74 @@
+          doc.selection.copy();
+  }selectCopy();
+      
+      switch (targetChannel) {
```
Since you're calling this function for each layer, you can skip the switch / case step!  Try this instead of the whole switch tree:

```
doc.activeLayer = doc.artLayers.getByName(targetChannel);
doc.paste();
```

2 lines should be able to do the job of 21!

Kaktug commented on split_to_CMYK.jsx:
```diff
@@ -1 +1,74 @@
+//copies channels into corresponding layers
+function splitChannels(targetChannel) {
+    
+    doc.activeLayer = doc.artLayers.getByName("Background");
```
What's the intent of this line here?  Looks like you're setting the 'highlighted' layer to the Background layer but it doesn't look like you're transforming it or using it as an index.

AlexBeers3 replied:
Good question.  So in order to copy any channel, the layer that has the art needs to be selected. eg the background layer otherwise if any of the empty layers are selected it throws an error that the layer is empty.  Therefore I need to make sure the background layer is selected in each instance of the splitChannels function or it will try to copy an empty layer.  Make sense?

Kaktug reviewed (changes requested):
Looking a lot cleaner (and DRYer)!  Left a few new comments / suggestions for you to look through.

Before you merge this PR we'll probably want to rewrite the branch history to remove all traces of that test.txt file.  I can help walk you through that (but we'll definitely want to do that while you're in the office since you had issues getting your text editor to pop open on your home computer).

AlexBeers3 committed:
simplify splitChannels function, add masks to each layer for atn file

AlexBeers3 committed:
add old jsx from home computer in order to pull from remote

AlexBeers3 committed:
merge home computer local with current remote

AlexBeers3 committed:
no new changes to split_to_cmyk

AlexBeers3 committed:
add two new jsx files clipart_me and shades_of_grey

AlexBeers3 committed:
add some code to clipart_me TESTING

AlexBeers3 committed:
add adj layers to code

AlexBeers3 committed:
action file that exports layers to files

AlexBeers3 commented:
Been playing around with some other scripts and added them to the repo.  
*clipart_me TESTING is just a file with a bunch of useful non-DOM code pulled from script listener for adj layers ect.  
*clipart_me.jsx is a new script that sets up a photo for trace in a clipart style two color
*shades_of_grey is like clipart_me but adj the photo to have simplified shades that can be traced
*export_tiffs.atn is an action that will export each layer individually with the file name and layer name following our naming conventions
*split_to_CMKY has been updated to make the adj layers through the script instead of running an external action.  
*The split_to_CMYK_adj_layers.atn is now redundant and can be removed at the same time we clean up / rewrite the branch history to remove all traces of that test.txt file as well.

AlexBeers3 committed:
clean up script listener functions to read better

AlexBeers3 committed:
add clipart_me_in_shades_of_grey.jsx

AlexBeers3 commented:
added combination script of shades of grey and clipart me "clipart_me_in_shades_of_grey" . Clipart me and shades of grey are now redundant
